model status refactor

barrettk opened the pull request:
- Defines "source of truth" function, that all other status checks depend on. All related functions determine the status based on the presence of a `bbi_config.json` file, migrating away from checking for `"Stop Time"` in a `.lst` file.

 - `get_model_status`, `check_nonmem_finished`, and `wait_for_nonmem` were all adjusted based on the above refactor. The exported methods and documentation are now **_much_** cleaner.

 - Groups all exported status related functions into the same documentation. Adds examples and additional details for certain model types.

 - Test adjustments.

closes https://github.com/metrumresearchgroup/bbr/issues/685

barrettk commented:
### Local Tests

```r
> devtools::test()
ℹ Testing bbr
✔ | F W  S  OK | Context
✔ |         21 | bbr exec functions                                                                                                                          
✔ |         13 | checking if models are up to date [1.3s]                                                                                                    
✔ |          9 | Collapse columns to string representation [1.1s]                                                                                            
✔ |         73 | Constructing config log from bbi_config.json [3.7s]                                                                                         
✔ |         43 | Copying model objects                                                                                                                       
✔ |          8 | copy-model-helpers                                                                                                                          
✔ |         34 | cov-cor                                                                                                                                     
✔ |         35 | Extract model paths from based_on fields [3.0s]                                                                                             
✔ |         34 | Test get_omega, get_sigma, and get_theta functions [2.4s]                                                                                   
✔ |         67 | Build paths from model object [1.6s]                                                                                                        
✔ |         22 | inherit-param-estimates [4.4s]                                                                                                              
✔ |         26 | initial-estimates [1.1s]                                                                                                                    
✔ |          7 | model_diff() comparing models                                                                                                               
✔ |         26 | Model status functions [2.0s]                                                                                                               
✔ |         19 | Test bbi summary on multiple models [1.0s]                                                                                                  
✔ |        108 | Test bbi summary functions [3.2s]                                                                                                           
✔ |         73 | Modify attributes of model object                                                                                                           
✔ |         25 | Testing function to create or read in model object                                                                                          
✔ |         35 | nm-file                                                                                                                                     
✔ |         38 | nm-join [3.0s]                                                                                                                              
✔ |         26 | nm-tables                                                                                                                                   
✔ |         30 | Test bbi batch parameter estimate functions [8.1s]                                                                                          
✔ |         11 | Test param_estimates functions [1.1s]                                                                                                       
✔ |         43 | test parsing labels for parameter table [1.6s]                                                                                              
✔ |         34 | testing print methods for bbi objects [8.1s]                                                                                                
✔ |          5 | read_bbi_path() helper function                                                                                                             
✔ |         50 | Reading NONMEM output files into R                                                                                                          
✔ |         66 | Constructing run log from model yaml [2.4s]                                                                                                 
✔ |         10 | submit_model(.dry_run=T)                                                                                                                    
✔ |         14 | submit_models(.dry_run=T)                                                                                                                   
✔ |         86 | Test creating summary logs [13.1s]                                                                                                          
✔ |          9 | Comparing tags between models                                                                                                               
✔ |         47 | test_threads(.dry_run=T) [8.3s]                                                                                                             
✔ |         21 | tweak-initial-estimates [1.1s]                                                                                                              
✔ |         16 | test-use-bbi [4.1s]                                                                                                                         
✔ |         34 | Utility functions for building args, etc.                                                                                                   
✔ |         43 | testing a composable workflow and running bbi [238.5s]                                                                                      
✔ |         72 | testing bootstrap functionality and running bbi [150.2s]                                                                                    

══ Results ══════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════
Duration: 471.7 s

[ FAIL 0 | WARN 0 | SKIP 0 | PASS 1333 ]
```

seth127 commented on R/utils.R:
```diff
@@ -565,234 +579,194 @@ bootstrap_is_cleaned_up <- function(.boot_run){
+  res_inc <- res$model_id[!res$finished]
+  if(length(res_inc) > 0 && length(res_inc) <= max_print){
+    mods_inc <- paste(res_inc, collapse = ", ")
+    message(glue("\n The following model(s) are incomplete: `{mods_inc}`"))
```
I think this (and the message on 723) should probably say "...are incomplete or have not yet been run"

This may be a nitpick, but to me this implies that those models are all _running_ but not yet complete (which is inaccurate). Especially because  `bbi_nonmem_model_status` (our "source of truth") does distinguish between those two statuses. Thoughts?

seth127 replied:
If you agree, this would require changing the expected strings in several tests.

seth127 reviewed (changes requested):
This is looking great overall, and seemed to work quite nicely in my ad-hoc testing.

I requested a few changes, almost all to docs or messages. Once those are done, I'm happy to approve and merge.

One final thought: consider moving all of the relevant code out of `R/utils.R` to a `R/model-status.R` file. 1) That `utils.R` file is really long and 2) we now have a `test-model-status.R` file (I think justifiably) so it probably makes sense to put that code in its own file too.

Nice work on this.

barrettk commented on R/utils.R:
```diff
@@ -565,234 +579,194 @@ bootstrap_is_cleaned_up <- function(.boot_run){
-  # look for model to be finished and then test output
-  model_finished <- if(file.exists(mod_path)){
-    read_lines(mod_path) %>% str_detect("Stop Time") %>% any()
+#' @describeIn model_status Returns messages indicating which model(s) have
```
Changed it to:
```r
#' @describeIn model_status Returns messages indicating which model(s) have
#'  finished executing and which are incomplete. Also invisibly returns a
#'  `data.frame` with a row for each model and a logical `finished` column.
```

barrettk commented:
Addressed comments:

<details closed>
<summary>Full Test Results</summary>

```r
> devtools::test()
ℹ Testing bbr
✔ | F W  S  OK | Context
✔ |         21 | bbr exec functions                                                                                                                          
✔ |         13 | checking if models are up to date [1.4s]                                                                                                    
✔ |          9 | Collapse columns to string representation [1.1s]                                                                                            
✔ |         73 | Constructing config log from bbi_config.json [3.3s]                                                                                         
✔ |         43 | Copying model objects                                                                                                                       
✔ |          8 | copy-model-helpers                                                                                                                          
✔ |         34 | cov-cor                                                                                                                                     
✔ |         35 | Extract model paths from based_on fields [3.7s]                                                                                             
✔ |         34 | Test get_omega, get_sigma, and get_theta functions [2.4s]                                                                                   
✔ |         67 | Build paths from model object [1.6s]                                                                                                        
✔ |         22 | inherit-param-estimates [4.0s]                                                                                                              
✔ |         26 | initial-estimates [1.3s]                                                                                                                    
✔ |          7 | model_diff() comparing models                                                                                                               
✔ |         26 | Model status functions [2.3s]                                                                                                               
✔ |         19 | Test bbi summary on multiple models                                                                                                         
✔ |        108 | Test bbi summary functions [3.3s]                                                                                                           
✔ |         73 | Modify attributes of model object                                                                                                           
✔ |         25 | Testing function to create or read in model object                                                                                          
✔ |         35 | nm-file                                                                                                                                     
✔ |         38 | nm-join [2.9s]                                                                                                                              
✔ |         26 | nm-tables                                                                                                                                   
✔ |         30 | Test bbi batch parameter estimate functions [8.2s]                                                                                          
✔ |         11 | Test param_estimates functions [1.1s]                                                                                                       
✔ |         43 | test parsing labels for parameter table [1.8s]                                                                                              
✔ |         34 | testing print methods for bbi objects [8.8s]                                                                                                
✔ |          5 | read_bbi_path() helper function                                                                                                             
✔ |         50 | Reading NONMEM output files into R                                                                                                          
✔ |         66 | Constructing run log from model yaml [2.3s]                                                                                                 
✔ |         10 | submit_model(.dry_run=T)                                                                                                                    
✔ |         14 | submit_models(.dry_run=T)                                                                                                                   
✔ |         86 | Test creating summary logs [12.7s]                                                                                                          
✔ |          9 | Comparing tags between models                                                                                                               
✔ |         47 | test_threads(.dry_run=T) [6.9s]                                                                                                             
✔ |         21 | tweak-initial-estimates [1.2s]                                                                                                              
✔ |         16 | test-use-bbi [4.1s]                                                                                                                         
✔ |         34 | Utility functions for building args, etc.                                                                                                   
✔ |         43 | testing a composable workflow and running bbi [233.1s]                                                                                      
✔ |         72 | testing bootstrap functionality and running bbi [151.3s]                                                                                    

══ Results ══════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════
Duration: 467.4 s

[ FAIL 0 | WARN 0 | SKIP 0 | PASS 1333 ]
```
</details>

seth127 reviewed (approved):
Looks great, thanks for the updates
